doc/design.md

```diff
@@ -0,0 +1,168 @@
+- History and depth QoS policies for each topic bridged
+- Remappings of topic, service, and action names
+- Optionally, it could be useful to override QoS settings for each entity being bridged.
+
+Not only should the bridge be configurable at an API level, but there should also be a way to configure it without having to change code (e.g. a passed in YAML or XML file).
+
+## Approach
+
+The proposed solution involves several nodes in a single bridge process.
+For each unique domain ID involved in the bridge process, a ROS context is created for that domain ID and a node is associated with the context.
+Therefore, we end up with one node per context per domain ID, all inside a single process.
+Bridged data will be shared between the nodes in the process so that it can be rebroadcast to a different domain.
+Nodes will be created as needed, based on the configuration of the bridge.
+
+For example, the following diagram illustrates a configuration where the "/chatter" topic from domain IDs `1` and `3` are both being bridged to domain ID `2`:
+
+![](topic_example.png)
+
+### API
+
+A C++ library with a public API is provided for users to call in their own process and extend as they like.
+The public API is expected to evolve over time, but at the very least users should be able to bridge ROS networks primitives
+
+For convenience, a standalone binary executable is also provided for easy integration into ROS systems.
+
+### Supporting generic types
+
+When bridging data, serialized data will be read from one domain and forwarded to another domain.
+As long as the bridge does not need access to the typed data, there is no need for deserialization and, therefore, C++ type support.
+In this way, we can handle arbitrary data coming from the middleware (as long as we load type support for the middleware).
+This is exactly how [rosbag2](https://github.com/ros2/rosbag2/tree/e4ce24cdfa7e24c6d2c025ecc38ab1157a0eecc8/rosbag2_transport) works; defining generic publishers and subscriptions.
+In fact, the generic publisher and subscription implementation is being moved to a common location that the domain bridge can leverage once available (see https://github.com/ros2/rclcpp/pull/1452).
+
+### QoS mapping
+
+TODO
+
+### Remapping
+
+TODO
+
+### Configuration
+
+The [YAML](https://yaml.org/) language was chosen as an easy way to externally configure the domain bridge.
+YAML is relatively simple to read for humans and computers alike and is commonly used in other places of the ROS ecosystem (e.g. ROS parameters).
+
+The bridge will look for three keys, `topics`, `services`, and `actions`, each of which contains a list of objects.
+The key for each object is the input name of the ROS entity (topic, service, or action name).
+For each ROS entity, you must specifiy the `type` of the interface, the domain ID where input data will come from (`from_domain`), and the domain ID for where to bridge (`to_domain`).
```

Comment:
In theory, I think we could drop the `type`. But in practice, in my limited understanding, we still need the type for loading some rmw support. I could dig deeper, but I'm inclined to wait and see how https://github.com/ros2/rclcpp/pull/1452 plays out, since that's where the `type` requirement is coming from.

Reply:
Sure.